mod_reforged/hooks/ai/tactical/behaviors/ai_switchto_ranged.nut

```diff
@@ -0,0 +1,36 @@
+::Reforged.HooksMod.hook("scripts/ai/tactical/behaviors/ai_switchto_ranged", function(q) {
+	// Prevent AI from swapping to ranged weapon when close to enemy melee units
+	// because dropping your Reach is a bad idea in this case
+	q.onEvaluate = @(__original) function( _entity )
+	{
+		this.m.WeaponToEquip = null;
+
+		local myTile = _entity.getTile();
+
+		foreach (enemy in this.getAgent().getKnownOpponents())
+		{
+			if (::MSU.isNull(enemy.Actor) || enemy.Actor.getMoraleState() == ::Const.MoraleState.Fleeing || enemy.Actor.isArmedWithRangedWeapon())
+			{
+				continue;
+			}
+
+			if (myTile.getDistanceTo(enemy.Tile) > ::Math.max(_entity.getIdealRange(), ::Const.AI.Behavior.SwitchToRangedMaxDangerDist))
+			{
+				continue;
+			}
+
+			if (enemy.Tile.getZoneOfControlCountOtherThan(enemy.Actor.getAlliedFactions()) >= ::Const.AI.Behavior.SwitchToRangedIgnoreDangerMinZones)
+			{
+				continue;
+			}
+
+			// If an enemy can reach us in 1 turn, don't swap to a ranged weapon
+			if (this.queryActorTurnsNearTarget(enemy.Actor, myTile, _entity).Turns <= 1)
```

Comment:
Does this include their ability to actually attack in the same turn?

Various enemie with throwing weapons also have Quickhands, which lets them instantly fix the bad position they found themselves in by switching to their bag melee weapon and attacking.

Reply:
No. There's a separate thing for that in that query result called `.TurnWithAttack`. However, the thing is that even if the enemy can't attack me this turn, maybe next turn he can acts before me and I won't be able to swap to a melee before that. So I find that it's probably better to just check if they can reach me in 1 turn.